### ISSUE TITLE
Provide an standard way to switch between prod and dev builds

### ISSUE THREAD

**Issue by izaera** [liferay/liferay-npm-build-tools#253]
Currently the making of dev/prod builds is left to the user and the bundler is agnostic in respect to that.

@yuchi proposes to have a way to tell the bundler which build you are doing and, with that, decide if we turn code like:

```javascript
'use strict';

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./cjs/react.production.min.js');
} else {
  module.exports = require('./cjs/react.development.js');
}
```

into:

```javascript
'use strict';

module.exports = require('./cjs/react.production.min.js');
```

for production, and leave untouched for dev.

That way, production builds always use production dependencies, while dev builds can branch depending on the value of the `process.env.NODE_ENV`.

**Comment by izaera** [comment]
Additionally we could have some way to configure the server in dev or prod mode that will set the value for the `process.env.NODE_ENV` variable.

**Comment by izaera** [comment]
@jbalsas Comments on this that defining `process.env.NODE_ENV` may pose a big risk because many libraries may be relying on the existence of `process` or `process.env` to do black magic and it could break everything.

**Comment by yuchi** [comment]
Typically those are used to check if the code is being run on Node or the browser, but they do more than that: https://github.com/MatthewNPM/is-node/blob/master/index.js

@jbalsas AFAIK all checks on `process.env.NODE_ENV` are things we “want” on production builds. Old-style modules loaded with browserify automatically get a `process` variable… so it should work…?

**Comment by wincent** [comment]
Would be nice to abstract this behind a `__DEV__` global which is not really a global. So that you can do `if (__DEV__)` in source and let the build take care of the details. Like Iván says, strip check entirely in prod, and in dev, turn it into something (could be a `process.env` check).

**Comment by yuchi** [comment]
@jbalsas totally agree. A deep investigation is definitely necessary.

@wincent This is not a concern for this project. It is a convention that developers can apply in theit own project, by specifying an additional babel-plugin. Liferay npm Build Tools could **promote** it.

**Comment by yuchi** [comment]
@wincent Also, keep in mind that in the React world `__DEV__` is just sugar for `process.env.NODE_ENV !== 'production'` AFAIK.

**Comment by izaera** [comment]
We can substitute `process.env.NODE_ENV` by `Liferay.ENV` or something similar to be sure we don't break anything.

Right now we are substituting `process.env.NODE_ENV` by `production` or `development` and leaving all references to `process.whatever` untouched so, substituting `process.env.NODE_ENV` by `Liferay.ENV` should work.

**Comment by yuchi** [comment]
@wincent Nope, I was wrong: https://github.com/facebook/metro/blob/a0b6ba0c68d4788816eb484f31cbc81e19ba425e/packages/metro/src/JSTransformer/worker/inline-plugin.js#L99-L117

**Comment by yuchi** [comment]
@izaera This is a surgically correct approach and could in fact be the right approach. Let me rewrite it so we are sure we are on the same page:


1. Introduce the concept of a `development` vs `production` mode in Liferay DXP, available as
   - `Liferay.ENV = 'development' | 'production'` on the frontend,
   - `PortalUtil.isProductionEnvironment()` on the JVM (so JSP, Java, FreeMarker…).
2. Introduce the concept of a `development` vs `production` target environment build parameter.
3. When npm Bundler is called for the `development` target environment:
   - all `process.env.NODE_ENV` occurrences are replaced with `Liferay.ENV`
   - all `__DEV__` occurrences are replaced with `Liferay.ENV !== 'production'`.
4. When npm Bundler is called for the `production` target environment:
   - all `process.env.NODE_ENV` occurrences are replaced with `"production"`
   - all `__DEV__` occurrences are replaced with `false`.

**Comment by wincent** [comment]
> Also, keep in mind that in the React world `__DEV__` is just sugar...

Yeah, I was asking for sugar (something easily and consistently applied everywhere without the developer having to think too much). You plan above sounds pretty much like that.

**Comment by wincent** [comment]
Going to do a mass-transfer of issues (including this one):

- **From:** https://github.com/liferay/liferay-js-toolkit/issues
- **To:** https://github.com/liferay/liferay-frontend-projects/issues

Note that many of the issues that we move may be stale, but we can close those once we've had time to review them. In the meantime, doing a bulk move makes is cheap and it will allow us to mark the liferay-js-toolkit repo as archived (read-only), which is an important final step in the consolidation process in order to avoid confusion.
